guides/keras_nlp/neural_machine_translation.py

```diff
@@ -0,0 +1,516 @@
+"""
+Title: English-to-Spanish translation with KerasNLP
+Author: [Abheesht Sharma](https://github.com/abheesht17/)
+Date created: 2022/05/26
+Last modified: 2022/05/26
+Description: Use KerasNLP to train a sequence-to-sequence Transformer model on the machine translation task.
+"""
+
+"""
+## Introduction
+
+KerasNLP provides building blocks for NLP (model layers, tokenizers, metrics, etc.) and
+makes it convenient to construct NLP pipelines.
+
+In this example, we'll use KerasNLP layers to build an encoder-decoder Transformer
+model, and train it on the English-to-Spanish machine translation task.
+
+This example is based on the [English-to-Spanish NMT example]
+(https://keras.io/examples/nlp/neural_machine_translation_with_transformer/) by
+[fchollet](https://twitter.com/fchollet). The original example is more low-level
+and implements layers from scratch, whereas this example uses KerasNLP to show
+some more advanced approaches, such as subword tokenization and using metrics
+to compute the quality of generated translations.
+
+You'll learn how to:
+
+- Tokenize text using `keras_nlp.tokenizers.WordPieceTokenizer`.
+- Implement a sequence-to-sequence Transformer model using KerasNLP's
+`keras_nlp.layers.TransformerEncoder`, `keras_nlp.layers.TransformerDecoder` and
+`keras_nlp.layers.TokenAndPositionEmbedding` layers, and train it.
+- Use `keras_nlp.utils.greedy_search` function to generate translations
+of unseen input sentences using the greedy decoding strategy!
+
+Don't worry if you aren't familiar with KerasNLP. This tutorial shows exactly
+how simple and easy it is to use it! So, what are you waiting for? Dive right in!
+"""
+
+"""
+## Setup
+
+Before we start implementing the pipeline, let's import all the libraries we need.
+"""
+
+"""shell
+pip install -q rouge-score
+pip install -q git+https://github.com/abheesht17/keras-nlp.git@rouge-l 
+"""
+
+import keras_nlp
+import numpy as np
+import pathlib
+import random
+import tensorflow as tf
+
+from tensorflow import keras
+from tensorflow_text.tools.wordpiece_vocab import bert_vocab_from_dataset as bert_vocab
+
+"""
+Let's also define our parameters/hyperparameters.
+"""
+
+BATCH_SIZE = 64
+EPOCHS = 1  # This should be at least 10 for convergence
+MAX_SEQUENCE_LENGTH = 40
+ENG_VOCAB_SIZE = 15000
+SPA_VOCAB_SIZE = 15000
+
+EMBED_DIM = 256
+INTERMEDIATE_DIM = 2048
+NUM_HEADS = 8
+
+"""
+## Downloading the Data
+
+We'll be working with an English-to-Spanish translation dataset
+provided by [Anki](https://www.manythings.org/anki/). Let's download it:
+"""
+
+text_file = keras.utils.get_file(
+    fname="spa-eng.zip",
+    origin="http://storage.googleapis.com/download.tensorflow.org/data/spa-eng.zip",
+    extract=True,
+)
+text_file = pathlib.Path(text_file).parent / "spa-eng" / "spa.txt"
+
+"""
+## Parsing the Data
```

Comment:
Section titles shouldn't capitalize all words (just the first one)

Reply:
Is there a reason for this? Curious. Because generally, we follow "Title Case" for headings, right?

Reply:
https://developers.google.com/style/capitalization#capitalization-in-titles-and-headings

Reply:
Interesting. Thanks!